Set color conditional

ggetz opened the pull request:
Added support for color conditionals. They are in the form:

```
color : {
   conditional : {
      'booleanExpression1` : `colorExpression1',
      'booleanExpression2` : `colorExpression2',
     // etc...
   },
   default : `defaultColorExpression`
}
```

We check each condition in order, and, if true, evaluate the corresponding color expression.

I added the file `ColorConditional.js` for the conditionals. It has both an `evaluate(feature)` and evaluate(feature, result)` function. I think we may only need the latter?

I deleted `ColorMapExpression` and `ColorRampExpression`.

I added tests for `ColorConditional` and for `Cesium3DTileStyle` and `Cesium3DTileset` to check if the style is properly applied.

I also updated the Sandcastle example. 

I'm having trouble with the `Change Style Using ID` and  `Change Style Using Height` buttons, because they generate the conditionals depending on the number of colors, and I was having trouble making the conditional JSON expression, because the Boolean expressions have to be strings: (ex. `'${height} < 1' : 'color("red")'` not `${height < 1 : 'color("red")'`). So that part of the Sandcastle example has not yet been updated.

pjcozzi commented:
> I think we may only need the latter?

I think we need both.  We should probably position this as a general "conditional", not specific to colors even if that is where the original use case came from.

ggetz commented:
> I think we need both. We should probably position this as a general "conditional", not specific to colors even if that is where the original use case came from.

OK, that should be pretty easy, Should I do that in this PR?

pjcozzi commented:
> Should I do that in this PR?

Yes, let me know if you run into any ideas.  I will also do the spec/schema for this tomorrow or Friday.

pjcozzi commented:
> I was having trouble making the conditional JSON expression, because the Boolean expressions have to be strings

What exactly is the problem?  Can't we just generate strings?

ggetz commented:
I set up the conditional like this:
```
var conditional  = {};
```
Then set up the conditional and color expression strings and use
```
conditional[conditionalExpression] = colorExpression;
```
but this outputs `{ ${height} < 1 : 'color("red")' }` not `{ '${height < 1' : 'color("red")' }`. Is there a way to make the "attribute" a string?

pjcozzi commented on Source/Scene/ColorConditional.js:
```diff
@@ -0,0 +1,130 @@
+        setRuntimeDefault(this);
+    }
+
+    defineProperties(ColorConditional.prototype, {
```
I would actually remove this for now until we expose the AST (probably not for the first pull request into `3d-tiles`).

Otherwise, this looks like a fast get/set, but it is just as bad as assigning a new style since it forces a recompile.

pjcozzi commented on Source/Scene/ColorConditional.js:
```diff
@@ -0,0 +1,130 @@
+        for (var cond in conditional) {
+            if (conditional.hasOwnProperty(cond)) {
+                var colorExpression = conditional[cond];
+                runtimeConditional.push({
```
Create a constructor function (local to this file) for this implicit type.  It will help with performance.

By the way, for performance tips, check this out when you have time: https://www.youtube.com/watch?v=XAqIpGU8ZZk

pjcozzi commented on Source/Scene/ColorConditional.js:
```diff
@@ -0,0 +1,130 @@
+        var defaultColor = this._runtimeDefault;
+        var conditional = this._runtimeConditional;
+        if (defined(conditional)) {
+            for (var i in conditional) {
```
Why not loop through this like a normal array?

pjcozzi commented on Source/Scene/ColorConditional.js:
```diff
@@ -0,0 +1,130 @@
+        var defaultColor = this._runtimeDefault;
+        var conditional = this._runtimeConditional;
+        if (defined(conditional)) {
+            for (var i in conditional) {
```
Same comment.

pjcozzi commented on Source/Scene/ColorConditional.js:
```diff
@@ -0,0 +1,130 @@
+        if (defined(conditional)) {
+            for (var i in conditional) {
+                if (conditional.hasOwnProperty(i)) {
+                    if (conditional[i].condition.evaluate(feature)) {
```
Store `conditional[i]` in a local.

pjcozzi commented:
Looks good, just those comments.

pjcozzi commented:
> but this outputs { ${height} < 1 : 'color("red")' } not { '${height < 1' : 'color("red")' }. Is there a way to make the "attribute" a string?

I still don't understand; is this with `JSON.stringify`?  The conditional is a string containing strings.  Is there an issue escaping them?

ggetz commented:
> I still don't understand; is this with JSON.stringify? The conditional is a string containing strings. Is there an issue escaping them?

I didn't use JSON.stringify- I will try it with that.

ggetz commented:
> I think we need both. We should probably position this as a general "conditional", not specific to colors even if that is where the original use case came from.

The default `default` value is currently the expression `color("#ffffff")`. Should we not have a default `default` value since we don't know what type this is supposed to evaluate to since it's being used in a general sense?

pjcozzi commented:
Let's remove `default` from the schema completely since a conditional that is just `'true'` at the end is the same thing.

ggetz commented:
In the case that someone doesn't provide a `'true'` condition, it will evaluate the expressions for any features that don't fall into the other conditionals as `undefined`. Is that behavior we want?

pjcozzi commented:
> Is that behavior we want?

Yes.

ggetz commented:
OK, I updated the name of the file to `ExpressionConditional` instead of `ColorConditional`, removed the default property, updated the tests, and updated the sandcastle example.

pjcozzi commented:
Looks great, just one rename request: throughout, rename `ExpressionConditional` to `ConditionalExpression`.

ggetz commented:
Updated.

pjcozzi commented:
:+1:
